packages/fuel-indexer-graphql/src/query.rs

```diff
@@ -0,0 +1,684 @@
+            OperationType::Subscription => Err(GraphqlError::OperationNotSupported(
+                "Subscription".to_string(),
+            )),
+        }
+    }
+}
+
+/// The type of selection that can be present in a user's operation.
+#[derive(Debug, Clone)]
+pub enum SelectionNode {
+    Scalar {
+        name: Name,
+        parent_entity: String,
+        alias: Option<Positioned<Name>>,
+    },
+    Object {
+        name: Name,
+        parent_entity: String,
+        alias: Option<Positioned<Name>>,
+        fields: Vec<SelectionNode>,
+        is_part_of_list: bool,
+        arguments: Vec<ParamType>,
+    },
+    List {
+        name: Name,
+        alias: Option<Positioned<Name>>,
+        node: Box<SelectionNode>,
+    },
+    RootNode {
+        name: Name,
+        alias: Option<Positioned<Name>>,
+        fields: Vec<SelectionNode>,
+        arguments: Vec<ParamType>,
+    },
+    PageInfo {
+        name: Name,
+        alias: Option<Positioned<Name>>,
+        fields: Vec<SelectionNode>,
+        parent_entity: String,
+    },
+    Connection {
+        name: Name,
+        alias: Option<Positioned<Name>>,
+        entity: String,
+        fields: Vec<SelectionNode>,
+        arguments: Vec<ParamType>,
+    },
+    Edge {
+        name: Name,
+        parent_entity: String,
+        alias: Option<Positioned<Name>>,
+        fields: Vec<SelectionNode>,
+        cursor: Box<Option<SelectionNode>>,
+        entity: String,
+    },
+}
+
+impl SelectionNode {
+    pub fn name(&self) -> String {
+        match &self {
+            SelectionNode::Scalar { name, .. } => name.to_string(),
+            SelectionNode::Object { name, .. } => name.to_string(),
+            SelectionNode::List { name, .. } => name.to_string(),
+            SelectionNode::RootNode { name, .. } => name.to_string(),
+            SelectionNode::PageInfo { name, .. } => name.to_string(),
+            SelectionNode::Connection { name, .. } => name.to_string(),
+            SelectionNode::Edge { name, .. } => name.to_string(),
+        }
+    }
+    /// Parses `SelectionNode` into a collection of strings that will
+    /// be used for generating a database query.
+    pub fn prepare(
+        &self,
+        schema: &ParsedGraphQLSchema,
+        db_type: &DbType,
+        dependency_graph: &mut DependencyGraph,
+    ) -> GraphqlResult<(Vec<String>, QueryParams)> {
+        let mut selection_set = vec![];
+        let mut query_params = QueryParams::default();
+
+        match db_type {
+            DbType::Postgres => {
+                let fqn = schema.fully_qualified_namespace();
+                match self {
+                    SelectionNode::Scalar {
+                        name,
+                        alias,
+                        parent_entity,
+                    } => {
+                        let field_name = alias
+                            .clone()
+                            .map_or(format!("'{}'", name), |a| format!("'{}'", a.node));
```

Comment:
In general, we would _love_ to avoid as much of this `format!` as possible. As mentioned in the other comments, if we could abstract this stuff behind informative structs or functions that would make it a bit more understandable and maintainable.